### PR TITLE
Allow custom settings in ClientRegistration.ClientSettings

### PR DESCRIPTION
**Pull request by pranavmanglik**
### Description
This pull request introduces an extensible `ClientSettings` configuration within `ClientRegistration`. Currently, `ClientRegistration` handles core OAuth2 properties, but adding provider-specific or advanced settings (such as custom PKCE requirements or extended client metadata) often requires modifying the core class or maintaining complex external maps.

By encapsulating these configurations in a dedicated, immutable `ClientSettings` object, we provide a cleaner API for future extensions without bloating the `ClientRegistration` root. This aligns with the framework's goals of modularity and extensibility for modern OAuth2/OIDC flows.

### Key Changes
* **Introduced `ClientSettings`**: A new configuration class held by `ClientRegistration` to store specialized client behavior settings.
* **Updated `ClientRegistration.Builder`**: Added the `.clientSettings(ClientSettings)` method to the builder pattern to allow for fluent configuration.
* **Enhanced Validation**: Integrated logic within `ClientRegistration.Builder` to ensure that provided `ClientSettings` (e.g., `requireProofKey`) are compatible with the selected `AuthorizationGrantType`.
* **Testing**: Added comprehensive unit tests in `ClientRegistrationTests` covering valid configurations, edge cases, and validation failures.

### Related Issues
Ref: #18863 Checklist
- [x] Fixed checkstyle issues using `./gradlew checkstyleMain`
- [x] Added unit tests in `ClientRegistrationTests`
- [x] All tests passed via `./gradlew :spring-security-oauth2-client:test`
- [ ] Documentation updated

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java [hunk 1]
```diff
@@ -755,14 +755,31 @@ public static final class ClientSettings implements Serializable {
...
+		}
+
+		@SuppressWarnings("unchecked")
+		public <T> @Nullable T getSetting(String name) {
+			Assert.hasText(name, "name cannot be empty");
+			return (T) this.settings.get(name);
+		}
+
+		@SuppressWarnings("unchecked")
+		public <T> T getSetting(String name, T defaultValue) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java]: Please remove this as I don't think providing a default is necessary.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java [hunk 2]
```diff
@@ -755,14 +755,31 @@ public static final class ClientSettings implements Serializable {
...
-		private ClientSettings() {
+		private final Map<String, Object> settings;
 
+		private ClientSettings(Map<String, Object> settings) {
+			this.settings = Collections.unmodifiableMap(new LinkedHashMap<>(settings));
 		}
 
 		public boolean isRequireProofKey() {
-			return this.requireProofKey;
+			return getSetting("settings.client.require-proof-key", true);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java]: Please extract `settings.client.require-proof-key` to a private constant

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java [hunk 3]
```diff
@@ -805,14 +823,35 @@ private Builder() {
...
+				this.settings.put(name, value);
+				return this;
+			}
+
+			/**
+			 * Sets the configuration settings.
+			 * @param settings the configuration settings
+			 * @return the {@link Builder} for further configuration
+			 */
+			public Builder settings(Map<String, Object> settings) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java]: Please change to `settings(Consumer<Map<String, Object>> settingsConsumer)`

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/ClientRegistrationTests.java [hunk 4]
```diff
@@ -751,4 +751,46 @@ private static <T> T getStaticValue(Field field, Class<T> clazz) {
 		}
 	}
 
+	@Test
+	void buildWhenScopesHaveInvalidCharactersThenThrowException() {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/ClientRegistrationTests.java:755]: All 3x tests are not related to the changes made to `ClientSettings`. Please remove and add applicable tests.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @pranavmanglik. Please see review comments.

Also, please rebase to latest on main. Thank you.

**Comment by pranavmanglik** [comment]
@jgrandja I think i have made the changes suggested by you, though i am not sure about these test cases, is there anything you suggest for that?

**Comment by jgrandja** [comment]
@pranavmanglik 

> is there anything you suggest for that?

Please take a look at these tests 

https://github.com/spring-projects/spring-security/blob/145579896fec5c31d9f0e3e81d41409a7f0baa12/oauth2/oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/settings/ClientSettingsTests.java#L30
